### PR TITLE
#567 Export data and structure of a table as csv

### PR DESCRIPTION
**Pull request by sworup**

### CODE REVIEW

**Comment by andreievg** [comment]
@sworup I am going to review this PR as a whole (rebased to master), to capture all of the changes made (sorry was late at getting this reviewed)

**Comment by sworup** [comment]
> @sworup I am going to review this PR as a whole (rebased to master), to capture all of the changes made (sorry was late at getting this reviewed)

@andreievg yes seems like the right thing to do as I have changed a lot from the previous base branch in this branch. Better to discard it and rebase this to master.

**Review by andreievg — CHANGES_REQUESTED** [review]
Hi Sworup,

On most part, code look pretty solid. I've left a few comments (had a look at contexts, hooks and queries). It looks like you've used most of the features of functional react, I wasn't sure why certain features were used (i.e. consumer vs just accessing state with context, and why fragments were used). I am sure the logic is sound, but it's hard to navigate the criss-cross dependencies in state between context, hooks and fragments, typically when the web of dependencies get's so intertwined a diagram is needed (to minimise the time someone new would spend to understanding the logic, or the developer themselves when coming back to make changes in code, or a PR reviewer).

We've been compiling such diagrams, they are in IRMIS/design/UI Workflows (they will eventually make their way into our documentation). One notable one (which was required since complexity of core system was starting to become overwhelming: https://app.diagrams.net/#G1EANnshpdQMk-uCXr7FMrRqIY-ESytIYV).

I've compiled a diagram of how I would implement the lookup table, this is just as an example (don't need to refactor to match my diagram), can you produce a digram with similar detail, to describe lookup table as they are now ? (this would help review the PR and for you to see improvements that can be done to make the code more digestible and to visualise the effects of state changes).

Here is the one I produced: https://app.diagrams.net/#G1OUI8fRnjTyvxBGYgrgpIyBoAk_55mKfu

lookupTable example diagram

Sorry I rushed in making it, so it's a little bit messy.

Btw, my opinionated personal note about React. It seems it's pretty easy to produce a solution in React where dependent elements of code are separated to an extend that it's very hard to visualise their dependencies and how and where states are changed and the effect the state change causes. I personally like the wrapper -> and hook approach, with minimal use of useEffect (one per hook or component if possible, and minimal dependencies in the useEffect). Basically a wrapper would use a hook and render a component or another wrapper, passing through hook results as props (yep, as props, there is a phobia of props use for some reason)

Also I feel that most of the time, context may reduce props propagation, but make components less re-usable and harder to understand (sometimes context are necessary though, for render optimisation or when stage changes need to affect sibling or parent components)

File: src/LookupTable/contexts/LookUpTableImportCsvContext.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  tableName: '',
+  submittable: false,
+  submitting: false,
+  errors: [],
+  success: false,
+}
+
+const LookUpTableImportCsvContext = React.createContext<{
+  state: LookUpTableImportCsvType
+  dispatch: React.Dispatch<any>
```
[the remark this answers is not in the document]
Comment (sworup) [reply src/LookupTable/contexts/LookUpTableImportCsvContext.tsx:17]: Resolved with 5cf6edfd and ad95cd6c

File: src/LookupTable/contexts/LookUpTableImportCsvReducer.tsx [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+const LookUpTableImportCsvReducer = (
+  state: LookUpTableImportCsvType,
+  action: LookUpTableImportCsvActionType
+): LookUpTableImportCsvType => {
+  switch (action.type) {
+    case LookUpTableImportCsvActions.UploadModalOpen:
+      return {
+        ...state,
+        uploadModalOpen: true,
```
[the remark this answers is not in the document]
Comment (sworup) [reply src/LookupTable/contexts/LookUpTableImportCsvReducer.tsx:15]: Resolved with 1b8e04dd

File: src/LookupTable/containers/LookupTableWrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  return (
+    <LookUpTableImportCsvProvider>
+      <Switch>
+        <Route exact path={[path, `${path}/import`]}>
+          <AllTableStructuresProvider>
+            <AllTableStructuresConsumer>
+              {({ getAllTableStructures }) => {
+                return (
+                  <LookupTableListPage
+                    onImportSuccess={() => getAllTableStructures()}
```
[the remark this answers is not in the document]
Comment (sworup) [reply src/LookupTable/containers/LookupTableWrapper.tsx:28]: Resolved with 0b717efd

File: src/LookupTable/graphql/getTableStructureById.query.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import { gql } from '@apollo/client'
+
+export default gql`
+  query($lookupTableID: Int!) {
```
[the remark this answers is not in the document]
Comment (sworup) [reply src/LookupTable/graphql/getTableStructureById.query.ts:4]: Resolved with 0b717efd

**Comment by andreievg** [comment]
going to merge this one as we need it for demo, @sworup can you transfer the comments I've made to a new issue please

**Review by andreievg — APPROVED** [review]

File: src/LookupTable/hooks/useGetSingleTable.hook.tsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+  useEffect(() => {
+    if (structure.id) {
+      setDynamicQuery(getDynamicSingleTable(structure))
+      getSingleTable()
+    }
+  }, [structure])
+
+  useEffect(() => {
+    const tableName = `${GQL_TABLE_NAME_PREFIX}${capitalizeFirstLetter(
```
[the remark this answers is not in the document]
Comment (sworup) [reply src/LookupTable/hooks/useGetSingleTable.hook.tsx:34]: Resolved with 2184f90f

File: src/LookupTable/components/ImportCsvModal.tsx [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+      <Modal.Header>
+        {!structure?.id ? 'Import Lookup-table' : `Import into Lookup-table: ${structure.label}`}
+      </Modal.Header>
+      <Modal.Content>
+        {submitting ? (
+          <Segment basic placeholder>
+            <Dimmer active inverted>
+              <Loader inverted>Submitting...</Loader>
+            </Dimmer>
+            <Image src="https://react.semantic-ui.com/images/wireframe/short-paragraph.png" />
```
[the remark this answers is not in the document]
Comment (sworup) [reply src/LookupTable/components/ImportCsvModal.tsx:108]: Resolved with abe3d502

**Pull request merged**